### PR TITLE
drivers: usb_dc_stm32: fix endpoint capabilities on STM32

### PR DESCRIPTION
**Pull request by aurel32**
This commit restores basic functionalities on STM32 for controller with bidirectional endpoints. I am not sure it does the right thing for unidirectional endpoint. For bidirectional endpoints, we might want to check that the type match, for example I am not sure you can have an isochronous and int type on the same endpoint.

### CODE REVIEW

**Comment by ydamigos** [comment]
> Now the reference manual actually says that it has:
> 1 bidirectional control endpoint0
> 5 IN endpoints
> 5 OUT endpoints

I checked the reference manual (RM0368) for  stm32f401xx and it mentions: 3 IN endpoints and 3 OUT endpoints. L4 and F7 series reference manuals mention 5 IN endpoints and 5 OUT endpoints. Regarding DTS files I used the descriptions in "Peripheral endpoints" section.

> do you know if they could actually work as bidirectional endpoints?
Each endpoint has its own control register (OTG_FS_DOEPCTL/OTG_FS_DIEPCTL or OTG_DOEPCTLx/OTG_DIEPCTLx). Reading the reference manuals, I understand that the application decides how to use the endpoints.

**Review by ydamigos — APPROVED** [review]
LGTM

**Comment by aurel32** [comment]
> I checked the reference manual (RM0368) for stm32f401xx and it mentions: 3 IN endpoints and 3 OUT endpoints. L4 and F7 series reference manuals mention 5 IN endpoints and 5 OUT endpoints. Regarding DTS files I used the descriptions in "Peripheral endpoints" section.

Oh, I didn't realized there was some segmentation at the USB controller level. The STM32F401 and the STM32F411 both have only 3 IN endpoints and 3 OUT endpoints, while the STM32F412, the STM32F413, and all the higher end ones have 5 IN endpoints and 5 OUT endpoints. It's probably something we want to fix, but that's out of scope of the current issue.

> Each endpoint has its own control register (OTG_FS_DOEPCTL/OTG_FS_DIEPCTL or OTG_DOEPCTLx/OTG_DIEPCTLx). Reading the reference manuals, I understand that the application decides how to use the endpoints.

Indeed, even if the user manual consider the IN and OUT endpoints as different endpoint, a given endpoint can both have the IN and OUT endpoints enabled. This is not like endpoints addresses 1 to 3 are OUT, while endpoints addresses 4 to 6 are IN. This is also not like an endpoint address can only be IN or OUT. So from the Zephy point of view they look like bidirectional.

Actually it looks like the current code already assumes them to be bidirectional given the following code:

```
/* Total in ep number = bidirectional ep number + in ep number */
#define NUM_IN_EP (CONFIG_USB_NUM_BIDIR_ENDPOINTS + \
                   CONFIG_USB_NUM_IN_ENDPOINTS)

/* Total out ep number = bidirectional ep number + out ep number */
#define NUM_OUT_EP (CONFIG_USB_NUM_BIDIR_ENDPOINTS + \
                    CONFIG_USB_NUM_OUT_ENDPOINTS)
/*
 * Total bidirectional ep number = bidirectional ep number + (out ep number +
 * in ep number) / 2.  Because out ep number = in ep number,
 * total bidirectional ep number = total out ep number or total in ep number
 */
#define NUM_BIDIR_EP NUM_OUT_EP

```

I think we should just consider all endpoints being bidirectional and just drop the extra complexity from the driver. Then `check_cap` can just compare `ep_idx` to `NUM_BIDIR_EP` independently of the direction.

**Review by loicpoulain — APPROVED** [review]

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -69,27 +69,10 @@
...
-#define NUM_OUT_EP (CONFIG_USB_NUM_BIDIR_ENDPOINTS + \
-		    CONFIG_USB_NUM_OUT_ENDPOINTS)
-/*
- * Total bidirectional ep number = bidirectional ep number + (out ep number +
- * in ep number) / 2.  Because out ep number = in ep number,
- * total bidirectional ep number = total out ep number or total in ep number
- */
-#define NUM_BIDIR_EP NUM_OUT_EP
+/* Total bidirectional endpoints, including endpoint 0 */
+#ifndef CONFIG_USB_NUM_BIDIR_ENDPOINTS
```
Comment (ydamigos) [inline drivers/usb/device/usb_dc_stm32.c]: We don't need this extra check. `num-bidir-endpoints` property is required in DT, so DT parser should handle it and raise an error if it is missing.

**Review by ydamigos — CHANGES_REQUESTED** [review]
I agree that the driver handles all endpoints as bidirectional and I am fine to describe all endpoints as bidirectional in DT in order to simplify driver's code (I should have done this in the first place).

Just a comment regarding changes.

**Comment by aurel32** [comment]
@ydamigos, thanks for the review. I have just pushed a new version which gets rid of this check.

**Review by ydamigos — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**